pkg/scenario/decompressor.go

```diff
@@ -0,0 +1,132 @@
+package scenario
+
+import (
+	"archive/zip"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	errs "github.com/ctfer-io/chall-manager/pkg/errors"
+	"github.com/pkg/errors"
+)
+
+const (
+	blockSize = 1 << 13 // arbitrary
+)
+
+// Decompressor handle the load of
+type Decompressor struct {
+	*Options
+}
+
+type Options struct {
+	MaxSize int64
+
+	currSize int64
+}
+
+// NewDecompressor constructs a fresh Decompressor.
+func NewDecompressor(opts *Options) *Decompressor {
+	if opts == nil {
+		opts = &Options{}
+	}
+	return &Decompressor{
+		Options: opts,
+	}
+}
+
+// Unzip extracts the content of the zip reader into cd.
+// It returns the directory it extracted into for Pulumi to use,
+// or an error if anything unexpected happens.
+func (dec *Decompressor) Unzip(r *zip.Reader, cd string) (string, error) {
+	outDir := ""
+	for _, f := range r.File {
+		if f.FileInfo().IsDir() {
+			continue
+		}
+		filePath, err := sanitizeArchivePath(cd, f.Name)
+		if err != nil {
+			return cd, &errs.ErrInternal{Sub: err}
+		}
+
+		// Save output directory i.e. the directory containing the Pulumi.yaml file,
+		// the scenario entrypoint.
+		base := filepath.Base(filePath)
+		if base == "Pulumi.yaml" || base == "Pulumi.yml" {
+			if outDir != "" {
+				return cd, errors.New("archive contain multiple Pulumi yaml/yml file, can't easily determine entrypoint")
+			}
+			outDir = filepath.Dir(filePath)
+		}
```

Comment:
This is required in order to handle 2 cases.

1. Flat zip
```
Pulumi.yaml
main
```

2. Directory zip
```
scenario/
├── Pulumi.yaml
└── main
```

Both encodes the same thing but are produced differently. In order for Pulumi to know in which directory to work, we try to find the Pulumi.yaml or Pulumi.yml file thus the directory it should be pointer to.